### ISSUE TITLE
Ensure discoverability of automatic garbage collection

### ISSUE THREAD

**Issue by matthieu-m** [rust-lang/cargo#13176]
### Problem

[Cache cleaning](https://blog.rust-lang.org/2023/12/11/cargo-cache-cleaning.html) is coming to Cargo, which I am really excited about.

Ideally, it should (once mature) be opt-out: active by default, with reasonable defaults, unless a user configures it. But how is a new user ever _supposed_ to know the feature exists and that it can be configured?

This may catch users unaware, wiping out their crate cache behind their back _before_ they embark on a journey in which they wished to revive an old project while sitting in a train, plane, in a cabin in the woods, etc...

### Proposed Solution

The GC feature should be _discoverable_, so that new users, even those who didn't read the documentation, or who forgot about it, are regularly reminded about its existence.

A potential solution would be to _remind_ users of the feature... whenever the feature kicks in. According to the [article](https://blog.rust-lang.org/2023/12/11/cargo-cache-cleaning.html#automatic-cleaning):

> When you run cargo, **once a day** it will inspect the last-use cache tracker, and determine if any cache elements have not been used in a while.

It would be very useful if at the very time automatic cleaning runs to determine what to clean, it would print:

 - Either a single line indicating there's nothing to clean.
 - Or, if it does clean anything, one line per crate/version that is cleaned.

Since this feedback is not time-sensitive, it would be fine -- should the cleaning run in the background -- to provide it on the _next_ run of `cargo` after cleaning completes instead. It would still be only once a day, and it would still make the feature discoverable by new users.

### Notes

[Bottom of the reddit comment chain which inspired this issue](https://www.reddit.com/r/rust/comments/18g1rg6/comment/kdd3588/?utm_source=share&utm_medium=web2x&context=3).

**Comment by epage** [comment]
> This may catch users unaware, wiping out their crate cache behind their back before they embark on a journey in which they wished to revive an old project while sitting in a train, plane, in a cabin in the woods, etc...

If the focus is on helping people when they go offline, would #13137 prevent people from running into that problem so its not needed?

As suggested, my biggest concern is bloated output.  People already complain about how noisy cargo is.

**Comment by cessen** [comment]
#13137 would certainly help alleviate the most common case of this.  But if I have a project versioned in git with e.g. feature branches that pull in additional dependencies, I don't think there's a good way for cargo to detect that without this feature getting significantly more complex.

And even if cargo went so far as to search git branches, what if I'm using a VCS other than git?  What if I keep some of my projects on an external USB drive?  What if I simply moved my project folders around while reorganizing my home directory?

The reality is that **Cargo doesn't (and can't) have control over the things that use its cache**, and so it fundamentally can't (reliably) know if any item in the cache is actually no longer used by something.  And trying to make the autoclean feature more sophisticated to handle more and more cases is never going to be enough, and would likely become a significant maintenance burden even if that were attempted.

So it would be both easier and (in some sense) more robust to just take the simple approach of ensuring that the user is informed and can opt out if needed/desired for their usage patterns.

> As suggested, my biggest concern is bloated output. People already complain about how noisy cargo is.

Yeah, noise can be an issue.  I would suggest just keeping the message short.  Maybe something along the lines of:

```
Cargo deleted 7 crates from its local cache that haven't been used in the last 3 months.
See <link to relevant manual page> for cargo garbage collection configuration options.
```

**Comment by weihanglo** [comment]
Alternatively, when `-Zgc` is disabled, we can emit something like what git does:

```
warning: There are too many unreachable loose objects; run 'git prune' to remove them
```

So that people are aware of the feature even when they intentionally disabled it.

**Comment by airstrike** [comment]
We could start by calling it "cache cleaning" rather than "garbage collection", which already means something else?

**Comment by cessen** [comment]
Yeah, that's a good point: garbage collection generally means cleaning stuff up that is provably unreachable/useless (garbage), whereas this is indeed a cache and (as outlined in this issue) its items may still have use.

**Comment by epage** [comment]
Note that naming is also being discussed on #12633.  We likely don't need it in two places and this is more about how to raise visibility, rather than what name will catch people's attention

**Comment by cessen** [comment]
Now that [cache cleaning is released](https://blog.rust-lang.org/2025/06/26/Rust-1.88.0/#cargo-automatic-cache-cleaning), I feel like this is really important to get in.  A simple version of this seems likely to be straightforward...?

Is the maintenance team open to a PR for this?

**Comment by epage** [comment]
Note that this is tagged `S-needs-design`. We've not fully settled on what this should look like.

**Comment by cessen** [comment]
Ah, got it.  In that case, what are the concrete steps to take so that a design can be settled on, and subsequently implemented?

I won't belabor the point, as it's already been discussed in this thread and elsewhere, but making cache cleaning discoverable (i.e. not an unknown unknown) for users who don't follow release notes seems important for avoiding the down sides of the feature for end users.  Time permitting, I would like to help push this to get in as soon as reasonably possible.

**Comment by epage** [comment]
So far the proposal is: always print a line of output when the GC is checked for cleaning.

As I mentioned, we already have problems with noisy output.   Two lines is too much for this imo.  One might be.

Now that I'm thinking of this more, some more challenges to consider when proposing a design include:
- all of the rest of Cargo's output might make this line unnoticeable, negating the benefit
- if r-a or another automation process is what triggers this, then the user will never see it

What can help stream line a proposal is to make it easy to review, including
- A balance of succinctness while fully communicating the idea. ("minimal yet complete")
- Comparing to alternative ideas
- Comparing to prior art within other tools

**Comment by cessen** [comment]
Thank you!  As time permits, I'll do my best to put together a proposal with these points in mind.

> Comparing to prior art within other tools

Was this done for the cache cleaning feature itself?  I ask not to be argumentative, but rather because I'm not familiar with which similar tools even do something like cargo's cache cleaning for comparison.  Therefore, such a list would save me time trying to track that down.

**Comment by epage** [comment]
We did look into how other tools deal with it. I'm on mobile and can't find the Issue atm but its either the tracking issue or one of the `-Zgc` tagged issues, possibly closed.

**Comment by cessen** [comment]
For my (and others') future reference, here is the tracking issue: #12633.

(Haven't read through it yet, but it at least looks like a good branching off point to discover the various places that discussion have taken place so far.)
